main/lsp/LSPTypechecker.cc

```diff
@@ -73,71 +73,32 @@ void LSPTypechecker::initialize(TaskQueue &queue, std::unique_ptr<core::GlobalSt
-
-        vector<ast::ParsedFile> indexed;
-        Timer timeit(config->logger, "initial_index");
-        ShowOperation op(*config, ShowOperation::Kind::Indexing);
-        vector<core::FileRef> inputFiles;
-        unique_ptr<const OwnedKeyValueStore> ownedKvstore = cache::ownIfUnchanged(*initialGS, move(kvstore));
-        {
-            Timer timeit(config->logger, "reIndexFromFileSystem");
-            inputFiles = pipeline::reserveFiles(initialGS, config->opts.inputFileNames);
-            indexed.resize(initialGS->filesUsed());
-
-            auto asts = hashing::Hashing::indexAndComputeFileHashes(*initialGS, config->opts, *config->logger,
-                                                                    absl::Span<core::FileRef>(inputFiles), workers,
-                                                                    ownedKvstore);
-            // asts are in fref order, but we (currently) don't index and compute file hashes for payload files, so
-            // vector index != FileRef ID. Fix that by slotting them into `indexed`.
-            for (auto &ast : asts) {
-                int id = ast.file.id();
-                ENFORCE_NO_TIMER(id < indexed.size());
-                indexed[id] = move(ast);
-            }
-        }
-
-        cache::maybeCacheGlobalStateAndFiles(OwnedKeyValueStore::abort(move(ownedKvstore)), config->opts, *initialGS,
-                                             workers, indexed);
-
-        ENFORCE_NO_TIMER(indexed.size() == initialGS->filesUsed());
-
-        updates.epoch = 0;
-        updates.typecheckingPath = TypecheckingPath::Slow;
-        updates.updatedFileIndexes = move(indexed);
-        updates.updatedGS = initialGS->deepCopy();
-
-        // Restore error queue, as initialGS will be used on the LSPLoop thread from now on.
-        initialGS->errorQueue = move(savedErrorQueue);
-    }
+    initialGS->trackUntyped = currentConfig.getClientConfig().enableHighlightUntyped;
 
     // We should always initialize with epoch 0.
-    this->initialized = true;
-    this->indexed = move(updates.updatedFileIndexes);
+    updates.epoch = 0;
+    updates.typecheckingPath = TypecheckingPath::Slow;
+    updates.updatedGS = std::move(initialGS);
+
     // Initialization typecheck is not cancelable.
     // TODO(jvilk): Make it preemptible.
-    auto committed = false;
+    SlowPathResult result;
     {
         const bool isIncremental = false;
         ErrorEpoch epoch(*errorReporter, updates.epoch, isIncremental, {});
         auto errorFlusher = make_shared<ErrorFlusherLSP>(updates.epoch, errorReporter);
-        committed = runSlowPath(move(updates), workers, errorFlusher, SlowPathMode::Init);
-        epoch.committed = committed;
+        result = runSlowPath(std::move(updates), std::move(kvstore), workers, errorFlusher, SlowPathMode::Init);
+        epoch.committed = true;
+        ENFORCE(std::holds_alternative<std::unique_ptr<core::GlobalState>>(result));
     }
-    ENFORCE(committed);
```

Comment:
WDYT about instead:

```c++
std::unique_ptr<GlobalState> state;
{
    ...
    auto result = runSlowPath(...);
    ENFORCE(std::holds_alternative<std::unique_ptr<core::GlobalState>>(result);
    state = std::move(std::get<...>(result));
}
ENFORCE(state);
```

...and then you'd have to use `state` in constructing `initTask`, below.

Reply:
Yeah, I like that better. Thanks!